File: sdk-python/src/workspace_sdk/services/process.py
"""
Process service for executing commands via gRPC
"""

from typing import Optional, Iterator, List

import grpc

from workspace_sdk.types import (
    CommandResult,
    RunCommandOptions,
    ProcessEvent,
    StdoutEvent,
    StderrEvent,
    ExitEvent,
    ErrorEvent,
)
from workspace_sdk.errors import convert_grpc_error
from workspace_sdk.proto.workspace.v1 import process_pb2, process_pb2_grpc


def _create_metadata(api_key: Optional[str]) -> List[tuple]:
    """Create gRPC metadata with auth token"""
    if api_key:
        return [("authorization", f"Bearer {api_key}")]
    return []


class ProcessService:
    """Sync service for executing commands in sandboxes via gRPC"""

    def __init__(
        self,
        stub: process_pb2_grpc.ProcessServiceStub,
        api_key: Optional[str],
        timeout: float,
    ):
        self._stub = stub
        self._api_key = api_key
        self._timeout = timeout

    def _metadata(self) -> List[tuple]:
        return _create_metadata(self._api_key)

    def run(
        self,
        sandbox_id: str,
        command: str,
        options: Optional[RunCommandOptions] = None,
    ) -> CommandResult:
        """Run a command and wait for completion"""
        req = process_pb2.RunCommandRequest(
            sandbox_id=sandbox_id,
            command=command,
            args=options.args if options and options.args else [],
            env=options.env if options and options.env else {},
        )
        if options and options.cwd:
            req.cwd = options.cwd
        if options and options.timeout:
            req.timeout_ms = options.timeout

        try:
            resp = self._stub.RunCommand(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            return CommandResult(
                exit_code=resp.result.exit_code,
                stdout=resp.result.stdout,
                stderr=resp.result.stderr,
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def run_stream(
        self,
        sandbox_id: str,
        command: str,
        options: Optional[RunCommandOptions] = None,
    ) -> Iterator[ProcessEvent]:
        """Run a command with streaming output"""
        req = process_pb2.RunCommandRequest(
            sandbox_id=sandbox_id,
            command=command,
            args=options.args if options and options.args else [],
            env=options.env if options and options.env else {},
        )
        if options and options.cwd:
            req.cwd = options.cwd
        if options and options.timeout:
            req.timeout_ms = options.timeout

        try:
            stream = self._stub.RunCommandStream(
                req, metadata=self._metadata(), timeout=self._timeout
            )
            for event in stream:
                parsed = self._parse_event(event)
                if parsed:
                    yield parsed
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def kill(
        self,
        sandbox_id: str,
        pid: int,
        signal: int = 15,
    ) -> None:
        """Kill a running process"""
        req = process_pb2.KillProcessRequest(
            sandbox_id=sandbox_id,
            pid=pid,
            signal=signal,
        )
        try:
            self._stub.KillProcess(
                req, metadata=self._metadata(), timeout=self._timeout
            )
        except grpc.RpcError as e:
            raise convert_grpc_error(e)

    def shell(
        self,
        sandbox_id: str,
        script: str,
        env: Optional[dict] = None,
    ) -> CommandResult:
        """Run a shell script using bash -c"""
        return self.run(
            sandbox_id,
            "bash",
            RunCommandOptions(args=["-c", script], env=env or {}),
        )

    def exec(
        self,
        sandbox_id: str,
        command: str,
        *args: str,
    ) -> str:
        """Execute a command and return stdout, raising on non-zero exit"""
        result = self.run(sandbox_id, command, RunCommandOptions(args=list(args)))
        if result.exit_code != 0:
            raise RuntimeError(
                f"Command failed with exit code {result.exit_code}: {result.stderr}"
            )
        return result.stdout

    def _parse_event(self, event) -> Optional[ProcessEvent]:
        """Parse proto ProcessEvent into SDK ProcessEvent"""
        which = event.WhichOneof("event")
        if which == "stdout":
            return StdoutEvent(type="stdout", data=event.stdout.data)
        elif which == "stderr":
            return StderrEvent(type="stderr", data=event.stderr.data)
        elif which == "exit":
            return ExitEvent(type="exit", code=event.exit.code)
        elif which == "error":
            return ErrorEvent(type="error", message=event.error.message)
        return None

File: sdk-typescript/src/services/process.ts
/**
 * Process service for executing commands via gRPC
 */

import * as grpc from '@grpc/grpc-js';
import { CommandResult, RunCommandOptions, ProcessEvent } from '../types/index.js';
import { ProcessServiceClient, createMetadata, promisifyUnary } from '../grpc.js';
import { convertGrpcError, ProcessError } from '../errors/index.js';

/**
 * Service for executing commands in sandboxes
 */
export class ProcessService {
  constructor(
    private readonly client: ProcessServiceClient,
    private readonly apiKey?: string
  ) {}

  private metadata(): grpc.Metadata {
    return createMetadata(this.apiKey);
  }

  /**
   * Run a command and wait for completion
   */
  async run(sandboxId: string, command: string, options: RunCommandOptions = {}): Promise<CommandResult> {
    try {
      const request: any = {
        sandboxId,
        command,
        args: options.args || [],
        env: options.env || {},
      };
      if (options.cwd) request.cwd = options.cwd;
      if (options.timeout) request.timeoutMs = options.timeout;

      const response = await promisifyUnary(
        this.client,
        this.client.runCommand,
        request,
        this.metadata()
      );

      return {
        exitCode: response.result.exitCode,
        stdout: response.result.stdout,
        stderr: response.result.stderr,
      };
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Run a command with streaming output
   */
  runStream(
    sandboxId: string,
    command: string,
    options: RunCommandOptions = {}
  ): AsyncIterable<ProcessEvent> {
    const client = this.client;
    const metadata = this.metadata();

    const request: any = {
      sandboxId,
      command,
      args: options.args || [],
      env: options.env || {},
    };
    if (options.cwd) request.cwd = options.cwd;
    if (options.timeout) request.timeoutMs = options.timeout;

    return {
      [Symbol.asyncIterator]: () => {
        const stream = client.runCommandStream(request, metadata);
        let done = false;
        let error: Error | null = null;

        stream.on('error', (err: grpc.ServiceError) => {
          error = convertGrpcError(err);
          done = true;
        });

        stream.on('end', () => {
          done = true;
        });

        return {
          next: (): Promise<IteratorResult<ProcessEvent>> => {
            return new Promise((resolve, reject) => {
              if (error) {
                reject(error);
                return;
              }

              if (done) {
                resolve({ value: undefined as any, done: true });
                return;
              }

              const onData = (data: any) => {
                stream.removeListener('data', onData);
                stream.removeListener('end', onEnd);
                stream.removeListener('error', onError);

                const event = this.parseEvent(data);
                if (event) {
                  resolve({ value: event, done: false });
                } else {
                  // Skip unknown events, get next
                  this.getNextEvent(stream, resolve, reject);
                }
              };

              const onEnd = () => {
                stream.removeListener('data', onData);
                stream.removeListener('end', onEnd);
                stream.removeListener('error', onError);
                done = true;
                resolve({ value: undefined as any, done: true });
              };

              const onError = (err: grpc.ServiceError) => {
                stream.removeListener('data', onData);
                stream.removeListener('end', onEnd);
                stream.removeListener('error', onError);
                reject(convertGrpcError(err));
              };

              stream.once('data', onData);
              stream.once('end', onEnd);
              stream.once('error', onError);
            });
          },
        };
      },
    };
  }

  private getNextEvent(
    stream: grpc.ClientReadableStream<any>,
    resolve: (value: IteratorResult<ProcessEvent>) => void,
    reject: (error: Error) => void
  ): void {
    const onData = (data: any) => {
      stream.removeListener('data', onData);
      stream.removeListener('end', onEnd);
      stream.removeListener('error', onError);

      const event = this.parseEvent(data);
      if (event) {
        resolve({ value: event, done: false });
      } else {
        this.getNextEvent(stream, resolve, reject);
      }
    };

    const onEnd = () => {
      stream.removeListener('data', onData);
      stream.removeListener('end', onEnd);
      stream.removeListener('error', onError);
      resolve({ value: undefined as any, done: true });
    };

    const onError = (err: grpc.ServiceError) => {
      stream.removeListener('data', onData);
      stream.removeListener('end', onEnd);
      stream.removeListener('error', onError);
      reject(convertGrpcError(err));
    };

    stream.once('data', onData);
    stream.once('end', onEnd);
    stream.once('error', onError);
  }

  /**
   * Kill a running process
   */
  async kill(sandboxId: string, pid: number, signal?: number): Promise<void> {
    try {
      const request: any = {
        sandboxId,
        pid,
      };
      if (signal !== undefined) request.signal = signal;

      await promisifyUnary(
        this.client,
        this.client.killProcess,
        request,
        this.metadata()
      );
    } catch (error) {
      throw convertGrpcError(error as grpc.ServiceError);
    }
  }

  /**
   * Convenience method: run a command and return stdout.
   * Throws ProcessError if the exit code is non-zero.
   */
  async exec(sandboxId: string, command: string, ...args: string[]): Promise<string> {
    const result = await this.run(sandboxId, command, { args });

    if (result.exitCode !== 0) {
      throw new ProcessError(
        sandboxId,
        command,
        `exit code ${result.exitCode}: ${result.stderr}`
      );
    }

    return result.stdout;
  }

  /**
   * Convenience method: run a shell command using bash -c
   */
  async shell(
    sandboxId: string,
    script: string,
    env?: Record<string, string>
  ): Promise<CommandResult> {
    return this.run(sandboxId, 'bash', {
      args: ['-c', script],
      env,
    });
  }

  private parseEvent(data: any): ProcessEvent | null {
    // Handle oneof field - proto-loader uses the field name directly
    if (data.stdout) {
      return { type: 'stdout', data: data.stdout.data };
    } else if (data.stderr) {
      return { type: 'stderr', data: data.stderr.data };
    } else if (data.exit) {
      return { type: 'exit', code: data.exit.code };
    } else if (data.error) {
      return { type: 'error', message: data.error.message };
    }
    return null;
  }
}
